File: transferir fragmentos archivos, hilos/grpc-transfer-project/src/main/java/fileshare/FileServerPorcentajeGrpc.java
package fileshare;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;

import java.io.FileOutputStream;
import java.io.IOException;

public class FileServerPorcentajeGrpc {

    private final int port;

    public FileServerPorcentajeGrpc(int port) {
        this.port = port;
    }

    public void start() throws IOException, InterruptedException {
        Server server = ServerBuilder.forPort(port)
                .addService(new FileServiceImpl())
                .build()
                .start();

        System.out.println("Servidor gRPC listo en puerto " + port + "...");
        server.awaitTermination();
    }

    public static void main(String[] args) throws IOException, InterruptedException {
        int port = 12345; // igual que tu versión con sockets
        new FileServerPorcentajeGrpc(port).start();
    }

    // Implementación del servicio
    static class FileServiceImpl extends FileServiceGrpc.FileServiceImplBase {

        @Override
        public StreamObserver<FileChunk> upload(StreamObserver<UploadStatus> responseObserver) {
            return new StreamObserver<FileChunk>() {

                private FileOutputStream fos;
                private String fileName = null;
                private long fileSize = 0;
                private long receivedBytes = 0;

                @Override
                public void onNext(FileChunk chunk) {
                    try {
                        // Primer mensaje: obtenemos metadatos y abrimos archivo
                        if (fos == null) {
                            fileName = chunk.getFileName();
                            if (fileName == null || fileName.isEmpty()) {
                                fileName = "archivo_recibido.dat";
                            }

                            fileSize = chunk.getFileSize();
                            fos = new FileOutputStream(fileName);

                            System.out.println("Iniciando recepción de archivo: " + fileName);
                            System.out.println("Tamaño declarado: " + fileSize + " bytes");
                        }

                        // Escribimos los datos
                        byte[] data = chunk.getData().toByteArray();
                        if (data.length > 0) {
                            fos.write(data);
                            receivedBytes += data.length;

                            if (fileSize > 0) {
                                int porcentaje = (int) ((receivedBytes * 100) / fileSize);
                                System.out.print("\rProgreso de recepción: " + porcentaje + "%");
                            } else {
                                System.out.print("\rRecibidos: " + receivedBytes + " bytes");
                            }
                        }

                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }

                @Override
                public void onError(Throwable t) {
                    t.printStackTrace();
                    try {
                        if (fos != null) fos.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }

                @Override
                public void onCompleted() {
                    try {
                        if (fos != null) fos.close();
                        System.out.println("\nRecepción completada de: " + fileName);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }

                    UploadStatus status = UploadStatus.newBuilder()
                            .setOk(true)
                            .setMessage("Archivo " + fileName + " recibido correctamente.")
                            .build();

                    responseObserver.onNext(status);
                    responseObserver.onCompleted();
                }
            };
        }
    }
}

File: practica_transferencia_final/08_grpc_streaming_porcentaje/FileServerPorcentajeGrpc.py
import grpc
from concurrent import futures
import file_service_pb2
import file_service_pb2_grpc

PORT = 50051


class FileServiceServicer(file_service_pb2_grpc.FileServiceServicer):

    def Upload(self, request_iterator, context):
        fos = None
        file_name = None
        file_size = 0
        received_bytes = 0

        try:
            for chunk in request_iterator:

                if fos is None:
                    file_name = chunk.file_name

                    if not file_name:
                        file_name = "archivo_recibido.dat"

                    file_size = chunk.file_size

                    fos = open(file_name, "wb")

                    print(f"Iniciando recepción de archivo: {file_name}")
                    print(f"Tamaño declarado: {file_size} bytes")

                data = chunk.data

                if data:
                    fos.write(data)
                    received_bytes += len(data)

                    if file_size > 0:
                        porcentaje = int((received_bytes * 100) / file_size)
                        print(f"\rProgreso de recepción: {porcentaje}%", end="")
                    else:
                        print(f"\rRecibidos: {received_bytes} bytes", end="")

            if fos:
                fos.close()

            print(f"\nRecepción completada de: {file_name}")

            return file_service_pb2.UploadStatus(
                ok=True,
                message=f"Archivo {file_name} recibido correctamente."
            )

        except Exception as e:
            if fos:
                fos.close()

            return file_service_pb2.UploadStatus(
                ok=False,
                message=str(e)
            )


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))

    file_service_pb2_grpc.add_FileServiceServicer_to_server(
        FileServiceServicer(),
        server
    )

    server.add_insecure_port(f"[::]:{PORT}")

    server.start()

    print(f"Servidor gRPC listo en puerto {PORT}...")

    server.wait_for_termination()


if __name__ == "__main__":
    serve()
